remove support for job-scoped auth tokens

ryanpetrello opened the pull request:
When Jobs and Adhoc Commands are launched, awx uses a job-scoped auth
token to dynamically fetch inventory via the awx REST API; this process
is complicated, hard to debug, and likely won't work going forward with
oauth2-based tokens in awx

see: https://github.com/ansible/awx/issues/21

ryanpetrello commented:
I took the nuclear option and also removed the `awxrest.py` inventory script entirely, because it's no longer necessary if we're not dynamically fetching the inventory file from the REST API.

AlanCoding commented on awx/main/models/inventory.py:
```diff
@@ -209,6 +209,70 @@ def get_group_children_map(self):
+                all_group = data.setdefault('all', dict())
+                smart_hosts_qs = self.hosts.all()
+                smart_hosts = list(smart_hosts_qs.values_list('name', flat=True))
+                all_group['hosts'] = smart_hosts
```
For both here, and the similar line a little lower, the more modern approach will be to put them in "ungrouped", and with our current Ansible support matrix, I think we can do this just about any time.

ryanpetrello replied:
👍 let's tackle that in another PR

AlanCoding commented on awx/main/tasks.py:
```diff
@@ -678,31 +677,13 @@ def should_use_proot(self, instance, **kwargs):
-                    'hostvars': True,
-                }).run(buff)
-                json_data = buff.getvalue().strip()
+                json_data = instance.inventory.get_script_data({}, hostvars=True, show_all=True)
```
This is so much more simple.

matburt replied:
👍

AlanCoding commented on awx/api/views.py:
```diff
@@ -2371,80 +2371,22 @@ class InventoryScriptView(RetrieveAPIView):
-
-        return Response(data)
+        return Response(obj.get_script_data(
+            data,
```
hold up, what's going on with this `data` parameter?

If query param "host" is specified, this is the same as passing `--host` to an inventory script. If this option is provided, the script should return a dictionary of the host variables, and only that. It looks like you have effectively decreased the indentation of the content that used to be nested under the `else:` here, which would do some kind of crazy mixing of the full script data with an individual host's data.

...and we should test this, of course. The ability to sanely test our script-generating method using the model's method is something that I've been looking forward to.

ryanpetrello replied:
Yep, you're totally right.  Good eye - I missed the indentation change here.  This should probably be more like:

```python
if hostname:
    host = get_object_or_404(obj.hosts, name=hostname, **dict(enabled=show_all))
    return Response(host.variables_dict)
return Response(obj.get_script_data(...))
```

AlanCoding commented on awx/main/models/inventory.py:
```diff
@@ -209,6 +209,70 @@ def get_group_children_map(self):
             group_children.add(from_group_id)
         return group_children_map
 
+    def get_script_data(self, data, hostvars=False, show_all=False):
```
yeah, so my vote would be to get rid of the `data` param here, and add back in the `host` param, which should have a default of `None`.

Here is an example of the options signature that I would expect this to match:

https://github.com/AlanCoding/Ansible-inventory-file-examples/blob/master/scripts/host_flag.py#L11

There are some other fancy things like --tree that `ansible-inventory` provide, and it would be the ideal model for the kwargs structure, although, of course, no need to add anything we're not using.

AlanCoding commented on awx/plugins/inventory/awxrest.py:
```diff
@@ -1,183 +0,0 @@
-            self.hostvars = bool(self.options.get('hostvars', False) or
-                                 os.getenv('INVENTORY_HOSTVARS', ''))
-            self.show_all = bool(self.options.get('show_all', False) or
-                                 os.getenv('INVENTORY_ALL', ''))
```
I think this shows:

```
self.show_all = bool(self.options.get('show_all', False) or
```

that the default value is False for calling awxrest.py.

See in tasks.py where you did

```
json_data = instance.inventory.get_script_data({}, hostvars=True, show_all=True)
```

I think you should set it to False, in order to not break the feature of host enablement. If the host isn't enabled, we don't want to run the playbook against it. Enablement is done by scripts, or can be toggled in the UI. Disabled hosts just can't have any jobs ran against them.

ryanpetrello replied:
Another good catch, it looks like the original behavior is *not* to specify `show_all`.

ryanpetrello replied:
My latest code fixes this; I tested by toggling enabled/disabled for a few hosts, and making sure they do/don't get used for playbook runs.

AlanCoding commented:
The two things in scope for PR merge:

 - kwargs structure for the script method, and assure that --host=hostname as queryparm works
 - all_hosts value

matburt reviewed (approved):
lgtm pending @AlanCoding's comments

ryanpetrello commented:
@AlanCoding I've uploaded some changes to address your comments - let me know what you think!

ryanpetrello commented:
@AlanCoding one hiccup with this approach is hostnames that contain unicode characters:

```
ERROR! Attempted to execute "/tmp/awx_17_7kOCSD/inventory" as inventory script: inventory data from /tmp/awx_17_7kOCSD/inventory contained characters that cannot be interpreted as UTF-8: 'utf8' codec can't decode byte 0xf1 in position 22: invalid continuation byte
```

I need to tinker and see what it takes to make this work.

ryanpetrello committed:
remove support for job-scoped auth tokens

When Jobs and Adhoc Commands are launched, awx uses a job-scoped auth
token to dynamically fetch inventory via the awx REST API; this process
is complicated, hard to debug, and likely won't work going forward with
oauth2-based tokens in awx

see: https://github.com/ansible/awx/issues/21

AlanCoding commented on awx/main/tasks.py:
```diff
@@ -678,31 +677,13 @@ def should_use_proot(self, instance, **kwargs):
-                json_data = buff.getvalue().strip()
-                f.write("#! /usr/bin/env python\nprint '''%s'''\n" % json_data)
+                json_data = json.dumps(instance.inventory.get_script_data(hostvars=True))
+                f.write("#! /usr/bin/env python\n# -*- coding: utf-8 -*-\nprint '''%s'''\n" % json_data)
```
Is this always a valid JSON representation? this is very interesting how you're just writing the data with JSON

AlanCoding replied:
d'oh, by this point is is a JSON string, not a dict as I was assuming

AlanCoding reviewed (approved):
ok, my comments are addressed

I think this is huge, as a step toward properly de-coupling the containers inside of the AWX deployment. Communication should go through the messaging system, and this closes a weird back route that we should avoid.

ryanpetrello commented:
Thanks @AlanCoding @matburt, I'll merge this once I fix the mocks in the busted tests.
